Complete user registration

prabeg-introcept opened the pull request:
- Add routes for user registration
- Create migration for users and departments table
- Configure 1:n relationship between Department and User model
- Seed multiple departments
- Add request validation for registration
- Add view for user-registration
- Store user and redirect to login with flash

prabeg-introcept committed:
Complete user registration

- Add routes for user registration
- Create migration for users and departments table
- Configure 1:n relationship between Department and User model
- Seed multiple departments
- Add request validation for registration
- Add view for user-registration
- Store user and redirect to login with flash

chandan-introcept commented on app/Http/Controllers/Auth/RegisterController.php:
```diff
@@ -0,0 +1,33 @@
+    public function register(RegisterUserRequest $request)
+    {
+        $user = User::create([
+            'username' => $request->username,
```
we can take only validated input like `$allDetails = $request->validated()` and instead of taking value from $request object each time we can take value from $allDetails array

chandan-introcept commented on resources/views/auth/register.blade.php:
```diff
@@ -0,0 +1,90 @@
+            @enderror
+        </div>
+        <div class="mb-3">
+            <label for="department" class="form-label">Department*: </label>
+            <select class="form-select" name="department_id">
+                <option selected value="1">Digital Marketing</option>
+                <option value="2">Project Management</option>
+                <option value="3">Design</option>
+                <option value="4">Development</option>
+                <option value="5">QA</option>
+                <option value="6">Customer Support</option>
```
Please-populate data from database because you have already created seeder for departments

chandan-introcept commented on database/migrations/2021_11_12_071755_add_column_department_id_to_users_table.php:
```diff
@@ -0,0 +1,36 @@
+     *
+     * @return void
+     */
+    public function down()
+    {
+        Schema::table('users', function (Blueprint $table) {
+            //
+        });
```
also write down logic

chandan-introcept commented on app/Http/Requests/RegisterUserRequest.php:
```diff
@@ -0,0 +1,54 @@
+        return [
+            'username' => 'required|string|max:255',
+            'email' => 'required|email|unique:App\Models\User,email|max:255',
+            'password' => 'required|string|min:8|regex:/^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{8,}$/',
```
we can follow this https://laravel.com/docs/8.x/validation#validating-passwords

prabeg-introcept committed:
Changes for PR review

- Use Laravel's Password rule object and confirmed validtion
- Fetch values from database for departments dropdown
- Store user data from object returned from validated method
- Use constants in validation rules
- Reorder migrations
- Typehint for relations and specify foreign keys
- Fetch departments from config for seeding
- Remove explicit mapping for user creation
- Render view for login from LoginController

ashish-shakya commented on app/Models/Department.php:
```diff
@@ -0,0 +1,16 @@
+
+    public function users()
+    {
+        return $this->hasMany(User::class);
```
Make a habit of define the foriegn key as second param.  The way you become clear about how relationship mapping is executed behind the scene.

So the better way would be 
        `return $this->hasMany(User::class, "user_id", "id");`

ashish-shakya commented on app/Models/User.php:
```diff
@@ -41,4 +43,14 @@ class User extends Authenticatable
+     */
+    public function department()
+    {
+        return $this->belongsTo(Department::class);
```
Make a habit of defining the foreign key as second param. The way you become clear about how relationship mapping is executed behind the scene.

ashish-shakya commented on app/Models/User.php:
```diff
@@ -41,4 +43,14 @@ class User extends Authenticatable
+     *
+     * @return BelongsTo
+     */
+    public function department()
```
Type hint the return type of the relationship as well

ashish-shakya commented on app/Models/Department.php:
```diff
@@ -0,0 +1,16 @@
+{
+    use HasFactory;
+
+    public function users()
```
Type hint the return type of the relationship as well

ashish-shakya commented on database/migrations/2014_10_12_000000_create_users_table.php:
```diff
@@ -15,11 +15,10 @@ public function up()
             $table->string('email')->unique();
-            $table->timestamp('email_verified_at')->nullable();
             $table->string('password');
-            $table->rememberToken();
```
why are `rememberToken` and `email_verified_at` field removed.
they do have specific purpose.
We Will need and may use later on in the course

ashish-shakya commented on database/migrations/2021_11_12_054445_create_departments_table.php:
```diff
@@ -0,0 +1,32 @@
+        Schema::create('departments', function (Blueprint $table) {
+            $table->id();
+            $table->string('department_name');
+            $table->timestamps();
```
since the table name is `departments`, field name can be written as `name` only rather than `department_name`

ashish-shakya commented on database/migrations/2021_11_12_071755_add_column_department_id_to_users_table.php:
```diff
@@ -0,0 +1,36 @@
+    public function down()
+    {
+        Schema::table('users', function (Blueprint $table) {
+            //
```
since yr creating a alter command add logic for artisan down command as well which reverts the previous action caused by migration

ashish-shakya commented on database/seeders/DepartmentSeeder.php:
```diff
@@ -0,0 +1,33 @@
+        foreach ($departments as $department)
+        {
+            DB::table('departments')->insert([
+                'department_name' => $department
```
Can be simply written as Department:create(["name"=>$department]) using elloquent. For this you have to define `Department` model.

ashish-shakya commented on resources/views/auth/register.blade.php:
```diff
@@ -0,0 +1,90 @@
+            <label for="department" class="form-label">Department*: </label>
+            <select class="form-select" name="department_id">
+                <option selected value="1">Digital Marketing</option>
+                <option value="2">Project Management</option>
```
pass these options from backend controller

ashish-shakya commented on routes/web.php:
```diff
@@ -13,6 +15,9 @@
-});
+Route::redirect('/', '/login');
+
+Route::view('/register', 'auth.register')->name('user.register');
```
for leveraging your involvement in laravel. Please take reference of `laravel-ui` package as well. It gives authentication module on the fly

ashish-shakya commented on database/seeders/DepartmentSeeder.php:
```diff
@@ -0,0 +1,33 @@
+    {
+        $departments = [
+            'Digital Marketing',
+            'Project Management',
```
better to maintain these department lists in a config file and retrieve it using `config()` helper function.

This way you can reuse these list anywhere in the application later on while developement

ashish-shakya commented on app/Http/Controllers/Auth/RegisterController.php:
```diff
@@ -0,0 +1,41 @@
+        $validatedUserData = $request->validated();
+
+        $user = User::create([
+            'username' => $validatedUserData->username,
```
see the usage of `$request->validated()` method.
it simply returns list of all data defined in RegisterUserRequest class.

Rather than mapping the validated one by one, it simply returns the set of all validated data

ashish-shakya commented on app/Http/Requests/RegisterUserRequest.php:
```diff
@@ -0,0 +1,63 @@
+        return [
+            'username.required' => 'Please enter your username',
+            'username.max' => 'Please enter less than 255 characters',
+            'email.required' => 'Please enter your email',
```
this is just to inform you that validation error messages are populated by default.

explicit definition is done only if we need to update the default validation message

ashish-shakya commented on database/migrations/2021_11_12_054445_create_departments_table.php:
```diff
@@ -0,0 +1,32 @@
+     */
+    public function down()
+    {
+        Schema::dropIfExists('departments');
```
not necessary, but it is good practise to define the table names as constants

ashish-shakya commented on app/Http/Requests/RegisterUserRequest.php:
```diff
@@ -0,0 +1,63 @@
+    {
+        return [
+            'username' => ['required', 'string', 'max:255'],
+            'email' => ['required', 'email', 'unique:App\Models\User,email', 'max:255'],
```
confirm the usage of 'unique:App\Models\User,email'

ashish-shakya replied:
i guess the syntax is `unique:<table_name>,<attribute_name>`
there is Rule::unique way of defining unique as well

ashish-shakya commented on routes/web.php:
```diff
@@ -13,6 +15,9 @@
+Route::get('/register', [RegisterController::class, 'index'])->name('user.register');
+Route::post('/register', [RegisterController::class, 'register']);
+
+Route::view('/login', 'auth.login')->name('user.login');
```
rather can returning view from routes. define a `LoginController` and return the login view  from controller method. This way you can resuse the controller for handling login action as well  later on

chandan-introcept commented on app/Constants/UsersTable.php:
```diff
@@ -0,0 +1,20 @@
+    const EMAIL_VERIFIED_AT = 'email_verified_at';
+    const PASSWORD = 'password';
+    const IS_ADMIN = 'is_admin';
+    const DEPARTMENT_ID = 'department_id';
```
i dont think these constants will help in readability.Is there any benifits  of creating this

chandan-introcept commented on app/Http/Requests/RegisterUserRequest.php:
```diff
@@ -0,0 +1,46 @@
+                    ->symbols()
+                    ->uncompromised()
+            ],
+            'confirmPassword' => ['required', 'string', 'same:password'],
```
we can simply use 'confirmed' in password field validation.

ashish-shakya commented on app/Http/Controllers/Auth/RegisterController.php:
```diff
@@ -0,0 +1,41 @@
+    /**
+     * @return Factory|View
+     */
+    public function index()
```
create method

chandan-introcept commented on app/Models/Department.php:
```diff
@@ -0,0 +1,22 @@
+     */
+    public function users(): HasMany
+    {
+        return $this->hasMany(User::class, 'department_id', 'id');
```
id is already default

ashish-shakya replied:
yaa chandan i let him add default values as well so that he could visualise what is going on behind the scene

chandan-introcept commented on app/Models/User.php:
```diff
@@ -41,4 +43,14 @@ class User extends Authenticatable
+     */
+    public function department(): BelongsTo
+    {
+        return $this->belongsTo(Department::class, 'department_id', 'id');
```
same here id is already default

chandan-introcept commented on config/db_values.php:
```diff
@@ -0,0 +1,12 @@
+<?php
```
please rename file name to meaningful name..like departments

chandan-introcept commented on database/migrations/2014_10_12_000005_create_users_table.php:
```diff
@@ -0,0 +1,44 @@
+     */
+    public function up()
+    {
+        Schema::create(DepartmentsTable::SCHEMA, function (Blueprint $table) {
```
??

ashish-shakya commented on database/seeders/DepartmentSeeder.php:
```diff
@@ -0,0 +1,27 @@
+
+        foreach ($departments as $department)
+        {
+            Department::create([
```
create => createOrUpdate, firstOrCreate()

chandan-introcept replied:
we can use updqateOrCreate in order to avoid redundancy when we run seeder multiple times

ashish-shakya commented on routes/web.php:
```diff
@@ -13,6 +15,9 @@
-});
+Route::redirect('/', '/login');
+
+Route::get('/register', [RegisterController::class, 'index'])->name('user.register');
```
Update to create method

ashish-shakya commented on app/Constants/DbTables.php:
```diff
@@ -0,0 +1,16 @@
+<?php
+
+
```
only single line must be given
